File: csharp/05_give_agency/Agent/Agent.Application/Agent.cs
namespace Agent.Application;

public class Agent()
{
    private readonly string _systemPrompt;
    private readonly IUserInput _input;
    private readonly IDisplay _display;
    private readonly ILanguageModel _model;

    public Agent(string systemPrompt, IUserInput input, ILanguageModel model, IDisplay display) : this()
    {
        _systemPrompt = systemPrompt;
        _input = input;
        _display = display;
        _model = model;
    }

    public void Run()
    {
        var context = new List<Message>
        {
            new("system", _systemPrompt)
        };

        while (true)
        {
            var userInput = _input.GetInput();
            if (string.IsNullOrWhiteSpace(userInput))
            {
                break;
            }
            var userMessage = new Message("user", userInput);
            _display.Show(userMessage);
            context.Add(userMessage);
            var answer = _model.Prompt(context);
            context.Add(answer);
            _display.Show(answer);

        }
    }
}

File: java/06_let_it_work/agent-application/src/main/java/com/workshop/agent/application/Agent.java
package com.workshop.agent.application;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class Agent {
    private final String systemPrompt;
    private final UserInput input;
    private final Display display;
    private final LanguageModel model;
    private final Tool tool;

    public Agent(String systemPrompt, UserInput input, LanguageModel model, Tool tool, Display display) {
        this.systemPrompt = systemPrompt;
        this.input = input;
        this.display = display;
        this.model = model;
        this.tool = tool;
    }

    public void run() {
        List<Message> context = new ArrayList<>();
        context.add(new Message("system", systemPrompt));

        while (true) {
            String userInput = input.getInput();
            if (userInput == null || userInput.isBlank()) {
                break;
            }
            Message userMessage = new Message("user", userInput);
            display.show(userMessage);
            context.add(userMessage);

            Message answer = model.prompt(context);
            display.show(answer);
            context.add(answer);

            Optional<String> toolResult = tool.parseAndExecute(answer.content());
            if (toolResult.isEmpty()) continue;

            Message toolResultMessage = new Message("user", toolResult.get());
            context.add(toolResultMessage);
            display.show(toolResultMessage);

            Message answerAfterTool = model.prompt(context);
            display.show(answerAfterTool);
            // not yet adding answerAfterTool to context
        }
    }
}
